### PR TITLE
Implement math:min and math:max

### LINKED ISSUE #727: Pour go's math functions into a `math:` module?
In the spirit of #576, should we also consider doing the same thing with some math functions?

This would benefit some people like me, who would like to open up terminals to do quick calculations without having to also type in an `ipython` each time I want to do so.

### PR DESCRIPTION
**Pull request by krader1961**
Resolves #727

### CODE REVIEW

**Comment by xiaq** [comment]
I prefer these functions to take variadic argument, in analogy with arithmetic functions like `+`.

I prefer these functions to throw an error when given no arguments. Elvish may get a Scheme-like numerical tower in future, and in such a system max and min should require at least one argument ([Racket's max](https://docs.racket-lang.org/reference/generic-numbers.html#%28def._%28%28quote._~23~25kernel%29._max%29%29) for example), which is is a consequence of two things:

* Functions that work for both inexact and exact numbers should only return inexact numbers when some of the arguments are inexact.

* +Inf and -Inf do not exist in the set of exact numbers.

**Comment by krader1961** [comment]
> I prefer these functions to take variadic argument, ...

That is my preference as well. However, it is not compatible with the `type Inputs` idiom that makes it easy to handle CLI args or input from the value stream. Consider this example:

```
> { put 1; put 2 } | +
▶ (float64 0)
```

You have to do this instead:

```
> { put 1; put 2 } | + (all)
▶ (float64 3)
```

This behavioral difference between variadic and "input list" commands has bothered me for quite a while. I can't find the issue but the current "input list" behavior was predicated on protecting against empty wildcard expansion. For example, `a_cmd *` which would consume its byte/value input if the glob expanded to nothing. The argument for the current "input list" behavior was that example should be written as `a_cmd [*]` to guard against the glob/wildcard expression not matching any files by resulting in an empty list that would keep `a_cmd` from reading from its stdin.

There is issue #318 but that isn't the one I recall. There was a different issue that argued for `a_cmd [*]` rather than `a_cmd *` because the glob might expand to nothing, thus causing `a_cmd` to read from its stdin.

That argument seems no longer relevant since a wildcard (i.e., glob) will throw an exception if it does not include the `[nomatch-ok]` modifier. In other words, the user has to explicitly say they understand the risk of using a wildcard that expands to nothing.

Also, consider functions like `str:join` and `joins`. Why does `joins : a b` throw a "arity mismatch" while `{ put a; put b } | joins :` works as expected? Why doesn't `joins` or `str:join` also take a variadic argument?

**Comment by krader1961** [comment]
The question about the behavior of an empty sequence should be moved to issue #727. Or, perhaps, a new issue since the #727 has been resolved, more or less, regardless of this change.

**Comment by xiaq** [comment]
See #727 and #991.

**Comment by xiaq** [comment]
While #991 is open, my preference remains making `math:min` and `math:max` variadic functions.

**Comment by krader1961** [comment]
> While #991 is open, my preference remains making math:min and math:max variadic functions.

That has me scratching my head; i.e., perplexes me. How often would you do `math:min 3 2 1 0`? Using the variadic form makes it much harder to use these functions in a value pipeline. Which is likely to be the common case. Even when selecting the min from a list I would expect to write `math:min $list` rather than `math:min $@list`.

**Comment by krader1961** [comment]
Not to mention expecting this to work:

```
{ put 1; put -1 } | math:min
```

Rather than having to do this:

```
{ put 1; put -1 } | math:min (all)
```

**Comment by xiaq** [comment]
Taking the smaller of two numbers is common; `math:min $a $b` feels more intuitive.

There isn't a good reason other than familiarity for `min` to work either way; most languages have `min` taking either two or variadic arguments; [C++](http://www.cplusplus.com/reference/algorithm/min/), [JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/min), [R6RS](http://www.r6rs.org/final/html/r6rs/r6rs-Z-H-14.html#node_sec_11.7.4.3).

**Comment by krader1961** [comment]
Okay, variadic form it is. Please take another look.

File: pkg/eval/math/math.go [hunk 1]
```diff
@@ -285,6 +285,50 @@ import (
...
+// ```
+//
+// Outputs the maximum number in the arguments. If there are no arguments
+// an exception is thrown. If any number is NaN then NaN is output.
+//
+// Examples:
+//
+// ```elvish-transcript
+// ~> math:max
+// ▶ ?(WTF)
```
Comment (xiaq) [inline pkg/eval/math/math.go:301]: Replace this with an actual transcript of the exception.

File: pkg/eval/math/math.go [hunk 2]
```diff
@@ -285,6 +285,50 @@ import (
...
+// ```
+//
+// Outputs the minimum number in the arguments. If there are no arguments
+// an exception is thrown. If any number is NaN then NaN is output.
+//
+// Examples:
+//
+// ```elvish-transcript
+// ~> math:min
+// ▶ ?(WTF)
```
Comment (xiaq) [inline pkg/eval/math/math.go:323]: Replace this with an actual transcript of the exception.

File: pkg/eval/math/math.go [hunk 3]
```diff
@@ -481,3 +527,17 @@ func (opts *isInfOpts) SetDefaultOptions() { opts.Sign = 0 }
 func isInf(opts isInfOpts, arg float64) bool {
 	return math.IsInf(arg, opts.Sign)
 }
+
+func max(fm *eval.Frame, num float64, nums ...float64) float64 {
```
Comment (xiaq) [inline pkg/eval/math/math.go:531]: Remove the `Frame` argument.

File: pkg/eval/math/math.go (hunk 1 above)
Comment (krader1961) [reply pkg/eval/math/math.go:301]: Argh! Yep, I meant that as a placeholder until I confirmed the exact spelling of the exception. Fixed.

File: pkg/eval/math/math.go (hunk 3 above)
Comment (krader1961) [reply pkg/eval/math/math.go:531]: Yep, no longer needed.

**Pull request merged**